### PR TITLE
add mTLS configuration support for gxlb (net-lb-app-ext)

### PR DESCRIPTION
**Pull request by g-greatdevaks**
The PR introduces the support for mTLS configuration for gxlb (net-lb-app-ext).
mTLS for GXLB became GA recently, however, the Terraform resources still use Beta.

I acknowledge that I have:
- [X] Read the [contributing guide](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md)
- [X] Ran `terraform fmt` on all modified files
- [X] Regenerated the relevant README.md files using [`tools/tfdoc.py`](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md#fabric-tools)
- [ ] Made sure all relevant tests pass

### CODE REVIEW

**Comment by aurelienlegrand** [comment]
Thanks for the PR! I think it could be interesting to add a full mTLS example in the readme.md file.

**Comment by g-greatdevaks** [comment]
Yes, thanks for the suggestion, @aurelienlegrand.
I am working on the documentation and adding test cases.

File: modules/net-lb-app-ext/variables.tf [hunk 1]
```diff
@@ -88,6 +88,21 @@ variable "labels" {
   default     = {}
 }
 
+variable "mtls_policy_config" {
+  description = "Configuration for mTLS for the HTTPS Load Balancer."
+  type = object({
+    client_validation_mode            = optional(string, "CLIENT_VALIDATION_MODE_UNSPECIFIED")
```
Comment (ludoo) [inline modules/net-lb-app-ext/variables.tf:94]: if this is an enum with a finite set  of values, we usually check it via a validation rule on the variable

File: modules/net-lb-app-ext/variables.tf [hunk 2]
```diff
@@ -88,6 +88,21 @@ variable "labels" {
...
+    create_mtls_resources             = optional(bool, false)
+    intermediate_cas_pem_certificates = optional(list(string), [])
+    policy_name                       = optional(string)
+    server_tls_policy_location        = optional(string, "global")
+    trust_config_name                 = optional(string)
```
Comment (ludoo) [inline modules/net-lb-app-ext/variables.tf:99]: would it make sense to group the three trust config attributes under a `trust_config = optional(object({}))`?

File: modules/net-lb-app-ext/main.tf [hunk 3]
```diff
@@ -28,6 +28,12 @@ locals {
     [for k, v in google_compute_ssl_certificate.default : v.id],
     [for k, v in google_compute_managed_ssl_certificate.default : v.id]
   )
+  trust_config_name = "trust-config-${var.name}"
```
Comment (ludoo) [inline modules/net-lb-app-ext/main.tf:31]: we don't prefix names with the resource in modules, usually; the name itself is fine

File: modules/net-lb-app-ext/main.tf [hunk 4]
```diff
@@ -28,6 +28,12 @@ locals {
...
+  mtls_policy_name  = "mtls-policy-${var.name}"
+}
+
+data "google_project" "project" {
```
Comment (ludoo) [inline modules/net-lb-app-ext/main.tf:35]: we avoid data sources as it makes it hard to test modules, can we derive the number from some other id already in the module, or ask the user to pass it in?

File: modules/net-lb-app-ext/main.tf [hunk 5]
```diff
@@ -78,14 +84,66 @@ resource "google_compute_target_http_proxy" "default" {
...
+    var.mtls_policy_config.trust_config_name != null
+    ? var.mtls_policy_config.trust_config_name
+    : local.trust_config_name
+  )
+  description = var.description
+  location    = var.mtls_policy_config.trust_config_location
+  trust_stores {
+    dynamic "trust_anchors" {
+      for_each = var.mtls_policy_config.trust_anchors_pem_certificates
+
```
Comment (ludoo) [inline modules/net-lb-app-ext/main.tf:115]: no newline here, and below in the other block

File: modules/net-lb-app-ext/main.tf [hunk 6]
```diff
@@ -78,14 +84,66 @@ resource "google_compute_target_http_proxy" "default" {
...
+    ? var.mtls_policy_config.policy_name
+    : local.mtls_policy_name
+  )
+  description = var.description
+  location    = var.mtls_policy_config.server_tls_policy_location
+  allow_open  = "false" # To be kept as false for external HTTPS Load Balancer. The field only applies for Traffic Director policies.
+  mtls_policy {
+    client_validation_mode         = var.mtls_policy_config.client_validation_mode
+    client_validation_trust_config = "projects/${data.google_project.project.number}/locations/${var.mtls_policy_config.trust_config_location}/trustConfigs/${google_certificate_manager_trust_config.default.0.name}"
+  } # A required field for HTTP Load Balancers if mTLS is to be enforced. Defines a mechanism to provision peer validation certificates for peer to peer authentication i.e. mTLS.
```
Comment (ludoo) [inline modules/net-lb-app-ext/main.tf:148]: comments go on their own line

File: modules/net-lb-app-ext/main.tf [hunk 7]
```diff
@@ -78,14 +84,66 @@ resource "google_compute_target_http_proxy" "default" {
...
+    var.mtls_policy_config.policy_name != null
+    ? var.mtls_policy_config.policy_name
+    : local.mtls_policy_name
+  )
+  description = var.description
+  location    = var.mtls_policy_config.server_tls_policy_location
+  allow_open  = "false" # To be kept as false for external HTTPS Load Balancer. The field only applies for Traffic Director policies.
+  mtls_policy {
+    client_validation_mode         = var.mtls_policy_config.client_validation_mode
+    client_validation_trust_config = "projects/${data.google_project.project.number}/locations/${var.mtls_policy_config.trust_config_location}/trustConfigs/${google_certificate_manager_trust_config.default.0.name}"
```
Comment (ludoo) [inline modules/net-lb-app-ext/main.tf:147]: this is better done in a local by joining tokens

**Review by ludoo — COMMENTED** [review]
thanks for this, I dropped a few initial comments

**Comment by ludoo** [comment]
Closing this as there's been no progress in 3 weeks. Feel free to reopen once suggestions have been taken into account.

**Pull request closed without merging**